### ISSUE TITLE
mate-dictionary segfault when run on gcc5 system w/ gtk3.17

### ISSUE THREAD

**Issue by lukefromdc** [mate-desktop/mate-utils#121]
An attempt to run mate-dictionary from any mate-utils package built with gtk3 (all I've tested) from either before or after Debian Unstable's gcc 5 transition errors out when run on a current, post-gcc5 Debian Unstable system. Worked fine earlier in the summer. Here's what I get in terminal:

(mate-dictionary:5263): GLib-GObject-WARNING **: The property GtkSettings:gtk-menu-images is deprecated and shouldn't be used anymore. It will be removed in a future version.

(mate-dictionary:5263): GLib-GObject-CRITICAL **: g_value_set_boxed: assertion 'G_VALUE_HOLDS_BOXED (value)' failed

(mate-dictionary:5263): GLib-GObject-WARNING **: /tmp/buildd/glib2.0-2.44.1/./gobject/gtype.c:4268: type id '0' is invalid

(mate-dictionary:5263): GLib-GObject-WARNING **: can't peek value table for type '<invalid>' which is not currently referenced
Segmentation fault

**Comment by monsta** [comment]
Not sure what you want us to do about it. Why not just rebuild it with gcc5?

**Comment by infirit** [comment]
I see you made it onto github luke :smile: 

From what I could gather GCC 5 now uses C99 by default which _could_ cause problems but it shouldn't. It is usually C++ that causes breakage.. Does a rebuild work?

I'll also start using Gcc5 on my gentoo machine and see how things go.

**Comment by flexiondotorg** [comment]
Ubuntu MATE 15.10 Beta 1 includes `mate-utils` 1.10.2 and was built with GCC5. `mate-dictionary` works correctly and doesn't segfault. I can't reproduce the issue being report.

**Comment by lukefromdc** [comment]
Rebuilding does not help or hurt: I tested builds from both before and after this
transition with the exact same results(I store debian packages from each build.

Interestingly, when I apply my modifications to network-manager-gnome, I HAVE to
specify C99 with older compilers, as that package normally builds with the -werror flag.
That causes the build to stop with the older compiler on 

GtkWidget *toplevel = gtk_widget_get_toplevel (menu);

as mixed declarations and code. Builds fine with C99, and no other MATE package has
been an issue with that. The C++ issue bit mate-system-monitor but required only a 
rebuild. You are probably right that this is the case with mate-dictionary as well, I 
figured as much. The transition in Debian came pretty much together and there 
were warnings to expect breakage. Even uglier is the refusal of EOM to build
at all, I opened another issue for that one.

Github did not complain about the faked "canvas" data, though the sign up GUI
was only partially responsive.

**Comment by lukefromdc** [comment]
I am building with gtk3, not gtk2 so that might be the difference. Alternaltely Ubuntu
might have done something differently in the C++ or gcc5 transition that made it
work.. I am on Debian Unstable with some Ubuntu packages.

**Comment by lukefromdc** [comment]
Rebuilding does not stop the segfaults, I tested both pre and post gcc5 builds
on a post gcc5 system with the same results

**Comment by lukefromdc** [comment]
On further testing, I have localized this to mate-dictionary's GUI. If called with -n for no-window it works and can print a definition of a word in terminal. Calling it without -n puts the error text in the terminal,  attempting to look up a word from the panel applet crashes the applet. Calling  mate-dictionary --gtk-debug=ALL got a bunch of terminal test about looking up icons followed by this:

Current icon themes UbuntuStudio_Legacy gnome hicolor mate Adwaita 

(mate-dictionary:27951): GLib-GObject-WARNING **: The property GtkSettings:gtk-menu-images is deprecated and shouldn't be used anymore. It will be removed in a future version.

(mate-dictionary:27951): Gtk-WARNING **: gtk_widget_create_pango_context ()) called without screen

(mate-dictionary:27951): GLib-GObject-CRITICAL **: g_value_set_boxed: assertion 'G_VALUE_HOLDS_BOXED (value)' failed

(mate-dictionary:27951): GLib-GObject-WARNING **: /tmp/buildd/glib2.0-2.44.1/./gobject/gtype.c:4268: type id '0' is invalid

(mate-dictionary:27951): GLib-GObject-WARNING **: can't peek value table for type '<invalid>' which is not currently referenced

Buried in that is a pango context called without screen, don't know if that's a clue or not

**Comment by lukefromdc** [comment]
UPDATE: this bug appears to be a bug in gtk 3.17 which is a prerelease version. It may not be a bug in the MATE code at all, as mate-dictionary works fine with gtk 3.16 on the exact same system. Unless this happens with the released gtk 3.18 in a few weeks it may be something that can be ignored.

**Comment by monsta** [comment]
No surprise :smile:
BTW, Debian doesn't have that version even in Unstable or Experimental, so I presume you've built it yourself?

**Comment by lukefromdc** [comment]
I've got over 60 locally built packages in use (compiz included) but got
gtk 3.17 from the Ubuntu GNOME PPA at

http://ppa.launchpad.net/ricotz/testing/ubuntu

Gtk3.18 is due out in weeks, Ubuntu 16.04 will probably use it, my guess
is everything in gtk3 builds of MATE 1.12 will need to be compatable with
it. 
One of the things I work with is a custom GTK theme, it generally pays
to at least test these against incoming versions of GTK so any breakage
can be fixed before the new version is released and reverting to the old
version becomes difficult. For me, gtk3.17 fixed some ugly sidebar 
issues, so I'm keeping it for normal use. Just so long as I know what 
made trouble with a dictionary that I don't use but which I do have to test
because it includes a panel applet and the panel is the focus of my 
MATE specific theme efforts.

**Comment by monsta** [comment]
@NiceandGently: I wonder if you can reproduce it in Fedora 23 (or maybe even Rawhide)...

**Comment by NiceandGently** [comment]
same in fedora 23 with gtk+-3.17.7, stacktrace at
https://dl.dropboxusercontent.com/u/49862637/Mate-desktop/Bugs/backtrace-mate-dict

**Comment by NiceandGently** [comment]
OK, th issue is caused by depcrecated style functions.
GtkWidget:visited-link-color and GtkWidget:link-color
Removing lines 23 and 24 from bluementa theme helps a lot :)
https://github.com/mate-desktop/mate-themes/blob/master/desktop-themes/BlueMenta/gtk-3.0/gtk-widgets.css#L23

**Comment by lukefromdc** [comment]
Removing the equivalent lines from my theme also stopped the segfaults. That's awful delicate coding in GTK 3 though where a theme entry-ANY theme entry-can cause a segfault.

**Comment by NiceandGently** [comment]
Yes, but this have to be fixed in code too for > gtk+-3.16
.... those entries are standard in most themes, i expect a lot of defaults with gtk+-3.18

**Comment by NiceandGently** [comment]
https://github.com/mate-desktop/mate-themes/commit/6b533752d741407500da4f856963194263bb4cfd

**Comment by monsta** [comment]
Looks like we found the culprit :smile:
@lukefromdc: I guess you can close this issue now?

**Comment by NiceandGently** [comment]
@monsta 
Why not trying to remove the depcrecated stuff from code?
Users who use Mint-X for example, or any other theme, will run in the same issue ;)

**Comment by monsta** [comment]
Hmm but if the theme doesn't use these two properties then apps won't crash, even w/o changing the code... or no?

**Comment by NiceandGently** [comment]
Yes, but only if the user use mate-themes.....
.....anyways, only a suggestion...
If you think this is OK for whlole linux world....than +1 for closing :)

**Comment by monsta** [comment]
Well, if we take Mint-X as an example, they're made for 3.10 and 3.14 (main Mint edition which is Ubuntu-based, and Linux Mint Debian Edition which is Debian-based). When we'll be making Mint 18 (based on Ubuntu 16.04), I'm sure @JosephMcc will look at your fixes :smile:

**Comment by JosephMcc** [comment]
So I received an email notification on this because I was mentioned. This should definitely be fixed in the code of the application if it can be. Just the fact that those lines exist in a theme should not cause a crash. Users get themes from a lot of places and not everyone follows all of the changing gtk version stuff.

**Comment by lukefromdc** [comment]
One potential headache comes from the fact that gtk3.14 themes mostly
work in all subsequent versions of gtk, not like the huge breakage between
say, 3.12 and 3.14. This is a good thing in and of itself, for instance my 
custom theme needs only ONCE change right now between 3.16 and 
3.17, two if this segfault is to be stopped. The 3.16 version also works
without issue in 3.14. 

Therefore, if this is not fixed either in application code or by GNOME in 
an unmaintained, deprecated function, there will be mystery segfaults in
mate-dictionary. Other applications will get this too, since GNOME is not
likely to fix this and some theme authors will never hear of it, every one
of these apps (ours included) need to fix this in the code.

I used the exact same logic in determining that Ubuntu may or may not
fix the gtk3.16 and later issues with transparency in compiz and fixed it
in my own code instead for my custom themable panel menus.

**Comment by lukefromdc** [comment]
Given later discussion about the issue remaining with random themes. we could
close it and post a "known issue" warning if gtk3.18 does not fix this, reopen it
at that time, or I could simply go through the code myself and see if I can find
it.   If I want to do that I should probably comment out, not delete. the offending
lines in my theme so I can enable them for "robust theme tolerance" testing.

**Comment by lukefromdc** [comment]
It won't but some users will have themes installed with the offending lines 
present and won't know how to edit them out.

**Comment by monsta** [comment]
> we could close it and post a "known issue" warning if gtk3.18 does not fix this

I'm sure they won't - removing some properties had never been considered a bug by them.
I agree that we should change the code - probably make it not to read these two properties at all if GTK+ > 3.16. Not sure how that might change the actual link colors in the app...

**Comment by monsta** [comment]
Dammit, same code is present in [MateAboutDialog](https://github.com/mate-desktop/mate-desktop/blob/master/libmate-desktop/mate-aboutdialog.c)...

**Comment by lukefromdc** [comment]
I just tested the mate-about dialog with the offending theme lines
re-enabled, no crash or segfault. Even has a link in it, no problem.
Wonder what is different?

**Comment by monsta** [comment]
I wonder if it's still relevant for us a year after...

**Comment by lukefromdc** [comment]
"Still relevant" would depend on whether or not anyone is still running unpatched GTK 3.16 themes with GTK 3.18. Such themes won't work at all with GTK 3.20 or 3.22, and have lots of things that will be ugly in GTK 3.18 due to lesser changes.

**Comment by raveit65** [comment]
This is caused by wrong settings in a theme, btw. if someone use a wrong theme gtk+ version.
I suggest to close it as we have have enough other reports with real problems.

**Comment by monsta** [comment]
Yes, I think all distros should have proper themes for their GTK+ versions by now.
